x-pack/plugins/security_solution/server/lib/detection_engine/rules/get_prepackaged_rules.ts

```diff
@@ -52,7 +53,43 @@ export const validateAllPrepackagedRules = (
   });
 };
 
+/**
+ * Retrieve and validate rules that were installed from Fleet as saved objects.
+ */
+export const getFleetInstalledRules = async (
+  client: RuleAssetSavedObjectsClient
+): Promise<AddPrepackagedRulesSchemaDecoded[]> => {
+  const fleetResponse = await client.all();
+  const fleetRules = fleetResponse.map(
+    // @ts-expect-error data is too loosely typed
+    (so) => so.attributes as AddPrepackagedRulesSchema
```

Comment:
I think you're shoe horning a bit here and this isn't the `validateAllPrepackagedRules` as it is written is really what you want maybe.

The `ts-expect-error` kind of gives it away as a bit as maybe something doesn't seem right.

If `validateAllPrepackagedRules` throws if the rule is not validated and was written when rules on a file system are a thing. Is that what you want? You want a throw here if it is not valid? Also it will say in the error message:

```ts
within the folder rules/prepackaged_rules 
```

I don't know if that's what you want either. The best bet would be to do something like this where you add a new function that is specific to checking and validating and throwing a specific error:

```ts
export const validateAllPrepackagedRulesFromSavedObjects = (
  rules: Array<Array<SavedObject<IRuleAssetSOAttributes & SavedObjectAttributes>>>
): AddPrepackagedRulesSchemaDecoded[] => {
  return rules.map((rule) => {
    const decoded = addPrepackagedRulesSchema.decode(rule);
    const checked = exactCheck(rule, decoded);

    const onLeft = (errors: t.Errors): AddPrepackagedRulesSchemaDecoded => {
      const ruleName = rule[0].attributes.name ? rule[0].attributes.name : '(rule name unknown)';
      const ruleId = rule[0].attributes.rule_id
        ? rule[0].attributes.rule_id
        : '(rule rule_id unknown)';
      throw new BadRequestError(
        `name: "${ruleName}", rule_id: "${ruleId}" within the saved object rules/prepackaged_rules ` +
          `is not a valid detection engine rule. Expect the system ` +
          `to not work with pre-packaged rules until this rule is fixed ` +
          `or the file is removed. Error is: ${formatErrors(
            errors
          ).join()}, Full rule contents are:\n${JSON.stringify(rule, null, 2)}`
      );
    };

    const onRight = (schema: AddPrepackagedRulesSchema): AddPrepackagedRulesSchemaDecoded => {
      return schema as AddPrepackagedRulesSchemaDecoded;
    };
    return pipe(checked, fold(onLeft, onRight));
  });
};

/**
 * Retrieve and validate rules that were installed from Fleet as saved objects.
 */
export const getFleetInstalledRules = async (
  client: RuleAssetSavedObjectsClient
): Promise<AddPrepackagedRulesSchemaDecoded[]> => {
  const fleetResponse = await client.all();
  const fleetRules = fleetResponse.map((so) => so.attributes);
  return validateAllPrepackagedRulesFromSavedObjects(fleetRules);
};
```

Then you can remove the `ts-expect-error` but also look at that bit of weirdness where I have two arrays?

```ts
      const ruleName = rule[0].attributes.name ? rule[0].attributes.name : '(rule name unknown)';
      const ruleId = rule[0].attributes.rule_id
        ? rule[0].attributes.rule_id
        : '(rule rule_id unknown)';
```

I don't think even that is 100% but it's closer to your intent of validating and then casting the data from a SO key/value generic data structure into a more non-generic data structure which has guaranteed fields such as the io-ts ones. I think maybe there just needs to be a loop over

Reply:
yeah, fair enough to make a separate function to validate here, so we don't assume any more than absolutely necessary about the incoming dynamically typed data.

I definitely need to do this check, I just mistakenly assumed that the `validateAllPrepackagedRules` would also be appropriate here.
```typescript
    const decoded = addPrepackagedRulesSchema.decode(rule);
    const checked = exactCheck(rule, decoded);
```

Reply:
41536f0e8a79029463d662cfde4fdc5c23528f93